contracts/src/v0.8/upkeeps/ERC20BalanceMonitor.sol

```diff
@@ -0,0 +1,289 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.4;
+
+import "../ConfirmedOwner.sol";
+import "../interfaces/KeeperCompatibleInterface.sol";
+import "@openzeppelin/contracts/security/Pausable.sol";
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+
+/**
+ * @title The ERC20BalanceMonitor contract.
+ * @notice A keeper-compatible contract that monitors and funds ERC20 tokens.
+ */
+contract ERC20BalanceMonitor is ConfirmedOwner, Pausable, KeeperCompatibleInterface {
+  IERC20 private ERC20Token;
+
+  uint256 private constant MIN_GAS_FOR_TRANSFER = 55_000;
+
+  event FundsAdded(uint256 amountAdded, uint256 newBalance, address sender);
+  event FundsWithdrawn(uint256 amountWithdrawn, address payee);
+  event TopUpSucceeded(address indexed topUpAddress);
+  event TopUpFailed(address indexed topUpAddress);
+  event WatchlistUpdated(address[] oldWatchlist, address[] newWatchlist);
+  event KeeperRegistryAddressUpdated(address oldAddress, address newAddress);
+  event ERC20TokenAddressUpdated(address oldAddress, address newAddress);
+  event MinWaitPeriodUpdated(uint256 oldMinWaitPeriod, uint256 newMinWaitPeriod);
+
+  error InvalidWatchList();
+  error OnlyKeeperRegistry();
+  error DuplicateAddress(address duplicate);
+
+  struct Target {
+    bool isActive;
+    uint96 minBalance;
+    uint96 topUpAmount;
+    uint56 lastTopUpTimestamp;
+  }
+
+  address private s_keeperRegistryAddress;
+  uint256 private s_minWaitPeriodSeconds;
+  address[] private s_watchList;
+  mapping(address => Target) internal s_targets;
+
+  /**
+   * @param erc20TokenAddress the ERC20 token address
+   * @param keeperRegistryAddress the address of the keeper registry contract
+   * @param minWaitPeriodSeconds the minimum wait period for addresses between funding
+   */
+  constructor(
+    address erc20TokenAddress,
+    address keeperRegistryAddress,
+    uint256 minWaitPeriodSeconds
+  ) ConfirmedOwner(msg.sender) {
+    setERC20TokenAddress(erc20TokenAddress);
+    setKeeperRegistryAddress(keeperRegistryAddress);
+    setMinWaitPeriodSeconds(minWaitPeriodSeconds);
+  }
+
+  /**
+   * @notice Sets the list of subscriptions to watch and their funding parameters.
+   * @param addresses the list of subscription ids to watch
+   * @param minBalances the minimum balances for each subscription
+   * @param topUpAmounts the amount to top up each subscription
+   */
+  function setWatchList(
+    address[] calldata addresses,
+    uint96[] calldata minBalances,
+    uint96[] calldata topUpAmounts
+  ) external onlyOwner {
+    if (addresses.length != minBalances.length || addresses.length != topUpAmounts.length) {
+      revert InvalidWatchList();
+    }
+    address[] memory oldWatchList = s_watchList;
+    for (uint256 idx = 0; idx < oldWatchList.length; idx++) {
+      s_targets[oldWatchList[idx]].isActive = false;
```

Comment:
any reason to not delete the whole struct? Will give you some gas refund if you do so

Reply:
So the thing is we could delete an item at ['0x1'] and get a gas refund, but if '0x1' is contained in the new watchlist it would be re-saved to storage, which would have the opposite effect from what we want.

Reply:
That should be fine. If you change a non zero slot to zero then to non zero again in same tx, gas costs will be handled automatically and you'll not be charged extra

Reply:
I see you decided not to change this, curious what were the reasons?

Reply:
Slipped my mind and then the audit occurred. My bad.